module/zfs/vdev_raidz.c

```diff
@@ -683,7 +683,7 @@ vdev_raidz_map_alloc_expanded(zio_t *zio,
 			 * rangelock, which is held exclusively while the
 			 * copy is in progress.
 			 */
-			if (rc->rc_size != 0 &&
+			if (!use_scratch && rc->rc_size != 0 &&
```

Comment:
The shadow location is only used for writes, and I don't think we do writes when `use_scratch` is set.  If we did do writes when `use_scratch` is set, then I think we'd need to write to the scratch region instead (e.g. `shadow_offset -= VDEV_BOOT_SIZE`).  So I think that this change should not be needed for correctness.

However, the change might be good since the current code wouldn't work if we actually did writes while using the scratch region.  But I wanted to make sure I'm not missing anything.  It would also be good to add a comment here explaining why we "don't need" to do the shadow write if the scratch region is in use.  Maybe in the caller (`vdev_raidz_io_start()`), we can assert that if we are doing a write, then `use_scratch` is not set?

Reply:
OK having ZIL writes makes sense.  And most writes will be handled correctly, by writing to the scratch region.  However, to make that fully work, if we have a write to a row that was partially copied to the scratch region, I think we need to do a "shadow write" to the scratch region, for any already-copied sectors.

Reply:
Done.
Please check, that I implemented it correctly.